### PR TITLE
Resolve or exempt the remaining Clippy errors & warnings

### PR DESCRIPTION
**Pull request by dwijnand**
With these changes Clippy runs cleanly now, i.e returns no warnings or errors.

Happy to tweak any details here, these are just my opening ideas on how to resolve these.

### CODE REVIEW

**Comment by Eh2406** [comment]
Over all I really like the changes, thanks for doing this! It is an important reminder that clippy has difrant things to say on different systems. Two small nits,

1. where we have ` // exempted`, looks like a good opportunity to explain *why* this is exempted. Like you did with ` // current intentional incoherence`.
2. the new type `DepFingerprint` looks like a good opportunity for a dock string explaining what the different parts meen.

**Comment by dwijnand** [comment]
Thanks for the reviews, @Eh2406 and @oli-obk!

I've pushed some commits addressing @Eh2406's good advice (thanks for the nudge! - I got lazy 🙂). WDYT?

File: src/cargo/core/compiler/job.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@ trait FnBox<A, R> {
 }
 
 impl<A, R, F: FnOnce(A) -> R> FnBox<A, R> for F {
+    #[cfg_attr(feature = "cargo-clippy", allow(boxed_local))] // The use of Box is intentional here
```
[the remark this answers is not in the document]
Comment (dwijnand) [reply src/cargo/core/compiler/job.rs]: Yeah, found it: https://github.com/rust-lang-nursery/rust-clippy/issues/1123

**Comment by Eh2406** [comment]
Looks good to me. I was too lazy to take on the project at all. So thanks!

@bors r+

**Pull request merged**